### ISSUE TITLE
add alternative to replaceAll that allows action when does not match

### ISSUE THREAD

**Issue by djschny** [elastic/elasticsearch#19661]
Currently `replaceAll` returns the original string if the pattern supplied does not match. This may not be ideal for all situations. Consider the following example where if a string does not match we might want to put them all into a single bucket (ie. return a default value).

cc @nik9000 

```
DELETE test
PUT /test
{
  "mappings": {
    "doc": {
      "properties": {
        "foo": {
          "type": "text",
          "fielddata": true
        }
      }
    }
  }
}
PUT test/doc/1
{
  "foo": "1.0"
}
PUT test/doc/2
{
  "foo": "2.0"
}
PUT test/doc/3
{
  "foo": "2.1"
}
PUT test/doc/4
{
  "foo": "v5.0"
}
GET /test/_search
{
  "size": 0, 
  "aggs": {
    "breakdownByMajorVersion": {
      "terms": {
        "script": {
          "lang": "painless",
          "inline": "/(^[1-9]+[0-9]?)\\..*/.matcher(doc['foo'].value).replaceAll('$1')"
        }
      }
    }
  }
}
```

The above works great when docs match the pattern, but if they don't we get something like the following:

```
  "aggregations": {
    "breakdownByMajorVersion": {
      "doc_count_error_upper_bound": 0,
      "sum_other_doc_count": 0,
      "buckets": [
        {
          "key": "2",
          "doc_count": 2
        },
        {
          "key": "1",
          "doc_count": 1
        },
        {
          "key": "v5.0",
          "doc_count": 1
        }
      ]
    }
  }
```

**Comment by clintongormley** [comment]
that's the contract of `replaceAll` in pretty much any language that implements it.  If you want to use a default in case nothing matches then you need to wrap this in an `if` statement and implement that check yourself.

**Comment by nik9000** [comment]
> If you want to use a default in case nothing matches then you need to wrap this in an if statement and implement that check yourself.

@djschny and I talked about this - it'd be useful to have a builtin that does this. I've reached for it when writing reindex scripts and had to reimplement it and he's reached for it with search scripts. Given that painless is designed to make stuff convenient with Elasticsearch I figure we should just implement this. I don't think it should be called `replaceAll` though.
